src/main/java/hudson/remoting/AtmostOneThreadExecutor.java

```diff
@@ -71,10 +71,11 @@ public boolean isTerminated() {
 
     public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
         synchronized (q) {
-            long start = System.currentTimeMillis();
-            long end = start+unit.toMillis(timeout);
-            while (isAlive() && System.currentTimeMillis()<end) {
-                q.wait(end-System.currentTimeMillis());
+            long now = System.nanoTime();
+            long end = now + unit.toNanos(timeout);
+            while (isAlive() && (end - now > 0L)) {
```

Comment:
if `timeout == 0` then `end = now + 0`  so (`end - now > 0 == false`) so we never enter the loop and call `wait(0)`